dts: Add new DTS/binding parser

ulfalizer opened the pull request:
```
dts: Add new DTS/binding parser

Add a new DTS/binding parser to scripts/dts/ for generating
generated_dts_board.conf and generated_dts_board_unfixed.h.

The old code is kept to generate some deprecated defines, using the
--deprecated-only flag. It will be removed later.

The new parser is implemented in three files in scripts/dts/:

dtlib.py:
  A low-level .dts parsing library. This is similar to devicetree.py in
  the old code, but is a general robust DTS parser that doesn't rely on
  preprocessing.

edtlib.py (e for extended):
  A library built on top of dtlib.py that brings together data from DTS
  files and bindings and creates Device instances with all the data for
  a device.

gen_defines.py:
  A script that uses edtlib.py to generate generated_dts_board.conf and
  generated_dts_board_unfixed.h. Corresponds to extract_dts_includes.py
  and the files in extract/ in the old code.

  Note that dtlib.py and edtlib.py are general libraries (similar to
  Kconfiglib) that could be used for other stuff besides gen_defines.py.

testdtlib.py:
  Test suite for dtlib.py. Can be run directly as a script.

testedtlib.py (uses test.dts and test-bindings/):
  Test suite for edtlib.py. Can be run directly as a script.

The test suites will be run automatically in CI.

The new code turns some things that were warnings (or not checked) in
the old code into errors, like missing properties specified with
'category: required' in the binding for the node.

The code includes lots of documentation and tries to give helpful error
messages instead of Python errors.

Co-authored-by: Kumar Gala <kumar.gala@linaro.org>
Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>
```

ulfalizer committed:
dts: Add new DTS/binding parser

Add a new DTS/binding parser to scripts/dts/ for generating
generated_dts_board.conf and generated_dts_board_unfixed.h.

The old code is kept to generate some deprecated defines, using the
--deprecated-only flag. It will be removed later.

The new parser is implemented in three files in scripts/dts/:

dtlib.py:
  A low-level .dts parsing library. This is similar to devicetree.py in
  the old code, but is a general robust DTS parser that doesn't rely on
  preprocessing.

edtlib.py (e for extended):
  A library built on top of dtlib.py that brings together data from DTS
  files and bindings and creates Device instances with all the data for
  a device.

gen_defines.py:
  A script that uses edtlib.py to generate generated_dts_board.conf and
  generated_dts_board_unfixed.h. Corresponds to extract_dts_includes.py
  and the files in extract/ in the old code.

testdtlib.py:
  Test suite for dtlib.py. Can be run directly as a script.

testedtlib.py (uses test.dts and test-bindings/):
  Test suite for edtlib.py. Can be run directly as a script.

The test suites will be run automatically in CI.

The new code turns some things that were warnings (or not checked) in
the old code into errors, like missing properties that are specified
with 'category: required' in the binding for the node.

The code includes lots of documentation and tries to give helpful error
messages instead of Python errors.

Co-authored-by: Kumar Gala <kumar.gala@linaro.org>
Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

galak commented on scripts/dts/extract_dts_includes.py:
```diff
@@ -533,7 +533,7 @@ def main():
     if args.include is not None:
         with open(args.include, 'w', encoding='utf-8') as f:
-            write_header(f, args.deprecate_only)
+            write_header(f, args.deprecated_only)
```
Pull this out into its own commit/PR.

ulfalizer replied:
Done

galak commented on cmake/dts.cmake:
```diff
@@ -175,6 +198,14 @@ if(SUPPORTS_DTS)
+  # Add the legacy identifiers
+  #
+
+  file(APPEND ${GENERATED_DTS_BOARD_UNFIXED_H}
```
what is this for?

ulfalizer replied:
Wrote about it on Slack. Seems the bsim tests (part of CI) do something funky where `generated_dts_board.h.deprecated` never gets generated. You get an #include error for it with the old approach.

galak replied:
Need to understand this better, I'm not in favor of this solution, and unclear why this makes 'bsim' happy.

ulfalizer replied:
@galak 
Are you looking into it?

The header doesn't show up, so it breaks otherwise.

galak reviewed (changes requested):
a few comments

ulfalizer committed:
scripts/dts: Rename --deprecate-only to --deprecated-only

--deprecate-only sounded like a command to "only deprecate (something)"
to me at first. --deprecated-only might make it clearer that it's about
only generating deprecated stuff.

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

ulfalizer committed:
scripts/dts: Add deprecation comments to old scripts

Makes it easier to understand what's going on.

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

pabigot reviewed (approved):
Review based on execution test only (didn't even look at the code).  

This worked nicely, including diagnosing a missing property, which cleans up one problem.

But it's missing the other half: it doesn't diagnose presence of property that is not in the yaml, such as a misspelled optional property.

That lack isn't worth blocking this, but if necessary I'll create an issue for it.

ulfalizer commented:
> But it's missing the other half: it doesn't diagnose presence of property that is not in the yaml, such as a misspelled optional property.
> 
> That lack isn't worth blocking this, but if necessary I'll create an issue for it.

Would be easy to add later, if we want to require all properties to be listed in the binding. Suspect we might have lots of nodes that don't follow that now.

pabigot commented:
> Would be easy to add later, if we want to require all properties to be listed in the binding. Suspect we might have lots of nodes that don't follow that now.

Probably not, unless this implementation (unlike the current one) generates defines for the extra properties.  If there are any, they can't be being used.

ulfalizer commented:
> > Would be easy to add later, if we want to require all properties to be listed in the binding. Suspect we might have lots of nodes that don't follow that now.
> 
> Probably not, unless this implementation (unlike the current one) generates defines for the extra properties. If there are any, they can't be being used.

Works like the old code. Thinking there might be a lot of unused stuff (or stuff that's undeclared because it's using some special path in the code). Maybe it's not that bad though...

b0661 commented on scripts/dts/edtlib.py:
```diff
@@ -0,0 +1,1626 @@
+
+        if len(paths) > 1:
+            _yaml_inc_error("multiple candidates for '{}' in !include: {}"
+                            .format(filename, ", ".join(paths)))
```
This prevents binding overload by application. An application (OOT) driver may need a special binding.
Using the first file in the list and reporting a warning would help in this case. An application can arrange to have it's bindings directory first in the list of directories searched.

ulfalizer replied:
The old code also gives an error here, so it's not a behavior change at least.

b0661 replied:
Yes it is like in the old code. Just a remark. I patched it in my project to have an easy way to develop the OOT drivers.

b0661 commented on scripts/dts/edtlib.py:
```diff
@@ -0,0 +1,1628 @@
+
+        bindings_dir:
+          Path to directory containing bindings, in YAML format. This directory
+          is recursively searched for .yaml files.
```
Should be a list of bindings directories to support applications that have their own bindings (additional to Zephyr).

galak commented on scripts/dts/gen_defines.py:
```diff
@@ -329,6 +301,23 @@ def dev_instance_aliases(dev):
+
+    out_comment("/chosen/{} ({})".format(prop_name, dev.path))
+    out("{}_BASE_ADDRESS".format(prefix), hex(dev.regs[0].addr))
+    out("SRAM_SIZE", dev.regs[0].size//1024)
```
is this missing `prefix`?

ulfalizer replied:
Oops... fixed

erwango reviewed (dismissed):
Can't build nucleo_f401re board (sample hello_world), due to missing symbol `CONFIG_RAM_SIZE`
It is defined in `arch/Kconfig`:
```
config SRAM_SIZE
	int "SRAM Size in kB"
	default $(dt_int_val,DT_SRAM_SIZE)
```
`DT_SRAM_SIZE` is not defined anymore

EDIT: Seems it has been replaced by:
```
#define DT_SRAM_SRAM_SIZE                           96
```
Cf fix in `scripts/dts/gen_defines.py`, line 318:
```
-    out("{}_SRAM_SIZE".format(prefix), dev.regs[0].size//1024)
+    out("{}_SIZE".format(prefix), dev.regs[0].size//1024)
```

ulfalizer commented:
> Can't build nucleo_f401re board (sample hello_world), due to missing symbol `CONFIG_RAM_SIZE`
> It is defined in `arch/Kconfig`:
> 
> ```
> config SRAM_SIZE
> 	int "SRAM Size in kB"
> 	default $(dt_int_val,DT_SRAM_SIZE)
> ```
> 
> `DT_SRAM_SIZE` is not defined anymore

Fixed by the latest force push. Thought I could make a small change without diffing `generated_dts_board.conf` before and after, but I couldn't. :P

erwango commented:
(Slack is broken to me so pasting my point here)
Would be nice to introduce nodes sorting at some step. 
From experience, this was useful in previous script when trying to debug and comparing 2 versions of generated output

ulfalizer commented:
@erwango 
Since Slack is dead:

The order in the conf and header file matches the order in the .dts file now, so it should be consistent at least. Would be pretty easy to change though, if you have some case where that order is bad.

erwango commented:
> The order in the conf and header file matches the order in the .dts file now, so it should be consistent at least. Would be pretty easy to change though, if you have some case where that order is bad.

This is not that I see, actually. For instance nucleo_f401re board, gpio nodes order:
`nucleo_f401re.dts_compiled`: A, B, C, D, E, H
`generated_dts_board_unfixed.h`: A, E, B, D, C, H (Changing at each build)

ulfalizer commented:
@erwango 
Try this patch and see if it fixes it. `set()` iteration order can vary between runs in Python.

```diff
diff --git a/scripts/dts/gen_defines.py b/scripts/dts/gen_defines.py
index 720c297615..9d8c43a9c7 100755
--- a/scripts/dts/gen_defines.py
+++ b/scripts/dts/gen_defines.py
@@ -72,7 +72,7 @@ def main():
             active_compats.update(dev.compats)
 
     out_comment("Active compatibles (mentioned in DTS + binding found)")
-    for compat in active_compats:
+    for compat in sorted(active_compats):
         #define DT_COMPAT_<COMPAT> 1
         out("COMPAT_{}".format(str2ident(compat)), 1)

```

ulfalizer commented:
@erwango 
What's your `python3 --version` btw? Dictionary iteration order didn't match insertion order before Python 3.6.

erwango commented:
```
$ python3 --version
Python 3.5.2
```

Tried your proposal, but it didn't work (didn't help sorting devices). But as I understand, this aims at sorting compatibles and not devices, so this is logical.

ulfalizer commented:
@erwango 
Yeah, the device-is-there flags use a `set()`, so they can move around between runs even on Python 3.6+, unless you sort them somehow.

The rest of it is consistent for me, so I'm guessing it's because you have Python 3.5.
